### PR TITLE
DRAFT: Option to print BTC transactions

### PR DESCRIPTION
**Pull request by webwarrior-ws**
Added `--print-transactions` option that for given Bitcoin
address prints all incoming transactions in reverse
chronological order.
Data printed: BTC amount, USD amount, date.

### CODE REVIEW

**Comment by knocte** [comment]
CI is broken @webwarrior-ws . Moving forward please don't create a PR yet if CI on your fork is not green yet.

File: src/GWallet.Frontend.Console/GWallet.Frontend.Console.fsproj [hunk 1]
```diff
@@ -83,14 +105,42 @@
...
-    </Reference>
     <Reference Include="SharpRaven">
       <HintPath>..\..\packages\SharpRaven.2.4.0\lib\net471\SharpRaven.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.CSharp" />
     <Reference Include="System.Configuration" />
     <Reference Include="System.Net.Http" />
+    <!-- FIXME: we should not need to reference ResultUtils at all when we can upgrade to F#v4.5 across the board (all CI lanes) -->
+    <Reference Include="ResultUtils">
+      <HintPath>..\..\packages\DotNetLightning.Kiss.1.1.2-date20220322-1031-git-5379324\lib\netstandard2.0\ResultUtils.dll</HintPath>
```
Comment (knocte) [inline src/GWallet.Frontend.Console/GWallet.Frontend.Console.fsproj:116]: @webwarrior-ws you're not using DNL in this PR so why do you add this? we told you to remove unnecessary changes from the PR please, review the PR changes yourself before telling us that you finished

File: src/GWallet.Backend/UtxoCoin/StratumClient.fs [hunk 2]
```diff
@@ -111,12 +137,89 @@ type public ElectrumServerReturningInternalErrorException(message: string, code:
                                                           originalRequest: string, originalResponse: string) =
     inherit ElectrumServerReturningErrorException(message, code, originalRequest, originalResponse)
 
-type StratumClient (jsonRpcClient: JsonRpcTcpClient) =
+module StratumRequestSerializer =
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/StratumClient.fs:140]: @webwarrior-ws same thing with this change, which seems to be a refactoring done in the LN branch (https://github.com/nblockchain/geewallet/commit/08fc3e53463d587dbeb2f1580fd5af4ff7b6a692) but not needed for this PR because this PR doesn't have tests

Comment (webwarrior-ws) [reply src/GWallet.Backend/UtxoCoin/StratumClient.fs:140]: But the module is used in `BlockchainScriptHashHistory` and `BlockchainTransactionGetVerbose` methods, which were not present before and so don't have "old" versions.

File: src/GWallet.Frontend.Console/GWallet.Frontend.Console.fsproj (hunk 1 above)
Comment (webwarrior-ws) [reply src/GWallet.Frontend.Console/GWallet.Frontend.Console.fsproj]: Removed reference to ResultUtils

File: src/GWallet.Backend/UtxoCoin/StratumClient.fs (hunk 2 above)
Comment (webwarrior-ws) [reply src/GWallet.Backend/UtxoCoin/StratumClient.fs:140]: @knocte

Comment (knocte) [reply src/GWallet.Backend/UtxoCoin/StratumClient.fs:140]: @webwarrior-ws then convert the BlockchainScriptHashHistory and BlockchainTransactionGetVerbose methods into old ones, it's very very very simple and the diff for the PR would be much smaller

Comment (webwarrior-ws) [reply src/GWallet.Backend/UtxoCoin/StratumClient.fs]: Converted the methods to "old style" and minimized changes in StratumClient

File: src/GWallet.Frontend.Console/packages.config [hunk 3]
```diff
@@ -1,7 +1,28 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="FSharp.Core" version="4.7.0" targetFramework="net461" />
+  <package id="FSharp.Core" version="4.7.0" targetFramework="net471" />
+  <package id="Microsoft.Bcl.AsyncInterfaces" version="5.0.0" targetFramework="net472" />
+  <package id="Microsoft.Extensions.Logging.Abstractions" version="1.0.2" targetFramework="net471" />
+  <package id="NBitcoin" version="6.0.17" targetFramework="net471" />
```
Comment (knocte) [inline src/GWallet.Frontend.Console/packages.config:6]: @webwarrior-ws this PR shouldn't need Frontend.Console adding NBitcoin as a dependency. I know you can figure out how to prevent this.

Comment (webwarrior-ws) [reply src/GWallet.Frontend.Console/packages.config]: Moved all logic to backend and removed all added dependencies.

Comment (knocte) [reply src/GWallet.Frontend.Console/packages.config]: > Moved all logic to backend and removed all added dependencies.

Thanks. Also can you rebase this PR to get CI green please.

Comment (webwarrior-ws) [reply src/GWallet.Frontend.Console/packages.config]: Done. CI is green.

File: src/GWallet.Backend/BTCTransactionPrinting.fs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                        None
+                let transactionInfo = NBitcoin.Transaction.Parse(transaction.Hex, NBitcoin.Network.Main)
+
+                let incomingOutputs =
+                    transactionInfo.Outputs 
+                    |> Seq.filter (fun output -> output.IsTo address)
+
+                if not (Seq.isEmpty incomingOutputs) then
+                    let amount = incomingOutputs |> Seq.sumBy (fun txOut -> txOut.Value)
+                    let dateTime = DateTime(1970, 1, 1) + TimeSpan.FromSeconds(float transaction.Time)
```
Comment (knocte) [inline src/GWallet.Backend/BTCTransactionPrinting.fs:60]: @webwarrior-ws what's this magic number?

File: src/GWallet.Backend/BTCTransactionPrinting.fs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                    let! bitcoinPrice = GetBitcoinPriceForDate dateTime
+                    Console.WriteLine(SPrintF1 "BTC amount: %A" amount)
+                    match bitcoinPrice with
+                    | Some price -> 
+                        let bitcoinAmount = amount.ToDecimal(NBitcoin.MoneyUnit.BTC) 
+                        Console.WriteLine(SPrintF1 "~USD amount: %s" ((bitcoinAmount * price).ToString("F2")))
+                    | None -> 
+                        Console.WriteLine("Could not get bitcoin price for the date")
```
Comment (knocte) [inline src/GWallet.Backend/BTCTransactionPrinting.fs:68]: @webwarrior-ws no more explanation? if this happens to any user, he would file a bug

File: src/GWallet.Frontend.Console/Program.fs [hunk 6]
```diff
@@ -480,5 +480,7 @@ let main argv =
         UpdateServersFile()
     | 1 when argv.[0] = "--update-servers-stats" ->
         UpdateServersStats()
+    | 2 when argv.[0] = "--print-transactions" ->
```
Comment (knocte) [inline src/GWallet.Frontend.Console/Program.fs:483]: @webwarrior-ws let's tweak this feature, if there's a number after the flag `--print-transactions` then you would only print that number of transactions (e.g. --print-transactions 3 would only print the last 3 transactions instead of continuing to print all).

File: src/GWallet.Backend/BTCTransactionPrinting.fs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+open System
+open System.Net
+
+open GWallet.Backend.FSharpUtil
+open GWallet.Backend.FSharpUtil.UwpHacks
+open GWallet.Backend.UtxoCoin
+
+
+module BTCTransactionPrinting =
```
Comment (knocte) [inline src/GWallet.Backend/BTCTransactionPrinting.fs:11]: @webwarrior-ws please rename to BtcTransactionPrinting and move it to UtxoCoin folder

File: src/GWallet.Backend/BTCTransactionPrinting.fs (hunk 4 above)
Comment (webwarrior-ws) [reply src/GWallet.Backend/BTCTransactionPrinting.fs:60]: Start of UNIX epoch

Comment (knocte) [reply src/GWallet.Backend/BTCTransactionPrinting.fs:60]: I know that; it's my way to say you have create a variable called unixEpoch, this is programming 101

File: src/GWallet.Frontend.Console/Program.fs (hunk 6 above)
Comment (webwarrior-ws) [reply src/GWallet.Frontend.Console/Program.fs]: done

File: src/GWallet.Backend/BTCTransactionPrinting.fs (hunk 7 above)
Comment (webwarrior-ws) [reply src/GWallet.Backend/BTCTransactionPrinting.fs]: done

File: src/GWallet.Backend/BTCTransactionPrinting.fs (hunk 4 above)
Comment (webwarrior-ws) [reply src/GWallet.Backend/BTCTransactionPrinting.fs]: done

**Comment by webwarrior-ws** [comment]
Fixed indentation in .fsproj file

**Comment by knocte** [comment]
@webwarrior-ws 2 last things to do here (while you're waiting for CI on the other task):
- Please rebase this PR with latest master in order to achieve green CI. After that:
- Please develop this small feature (in a 2nd commit, do not squash with 1st): when user writes a parameter for maxTransactionsCount, for example 2, check if the last 2 transactions have the same addresses as outputs (e.g. tx 1 outputs are 3: sends to address A, which belongs to wallet, and to address B and C, which don't belong to wallet; tx2 outputs are 3: sends to address A, which belongs to wallet, and to address B and D, which don't belong to the wallet), and if that is the case, put info about the TOTAL amount being sent, in tx1: A+B amount and percentage of each amount (e.g. 75% to A, 25% to B), and tx2: A+B amount and percentage of each amount (e.g. 65% to A, 35% to B).

File: src/GWallet.Frontend.Console/Program.fs [hunk 8]
```diff
@@ -480,5 +480,11 @@ let main argv =
         UpdateServersFile()
     | 1 when argv.[0] = "--update-servers-stats" ->
         UpdateServersStats()
+    | 2 | 3 when argv.[0] = "--print-transactions" ->
```
Comment (knocte) [inline src/GWallet.Frontend.Console/Program.fs:483]: @webwarrior-ws how are you testing this? this is wrong, don't you see from the 2 lines above that you should use `1 | 2` instead of `2 | 3`?

Comment (knocte) [reply src/GWallet.Frontend.Console/Program.fs:483]: @webwarrior-ws disregard the above message; there has been a misunderstanding, will explain later

File: src/GWallet.Frontend.Console/Program.fs [hunk 9]
```diff
@@ -480,5 +480,11 @@ let main argv =
...
+        // --print-transactions [maxTransactionsCount] <btcAddress>
```
Comment (knocte) [inline src/GWallet.Frontend.Console/Program.fs:484]: @webwarrior-ws it should be the other way around, the optional argument comes the last

**Comment by knocte** [comment]
I gave this a go today, I found 2 issues @webwarrior-ws , see this example output (where I cut addresses with `<snip/>` for privacy):

```
% mono src/GWallet.Frontend.Console/bin/Debug/GWallet.Frontend.Console.exe --print-transactions 2 39...<snip/>
BTC amount: 0.02800286
~USD amount: 478.88
date: ... UTC

BTC amount: 0.02321512
~USD amount: 391.20
date: ... UTC

Transactions with shared outputs:
Transaction with outputs to NBitcoin.TxOut, NBitcoin.TxOut
Date: ... PM UTC
Total BTC: 0.03634679
Sent 0.00134309 BTC to bc1qp...(snip) (3.70%)
Sent 0.00700074 BTC to bc1q1...(snip) (19.26%)

Transaction with outputs to NBitcoin.TxOut, NBitcoin.TxOut
Date: 8/12/2022 8:33:59 PM UTC
Total BTC: 0.03034462
Sent 0.00132585 BTC to bc1qp...(snip) (4.37%)
Sent 0.00580375 BTC to bc1q1...(snip) (19.13%)
```

The two issues are:
- The percentages are wrong, because their total is not 100%.
- Most importantly, address `39...<snip/>` is not included in the final summary titled `Transactions with shared outputs`, it should be included!

@webwarrior-ws just a reminder, this is a **low-priority** task, you should only work on this when you're waiting for CI for your high-priority task. Thanks Taras, good progress so far.

File: src/GWallet.Frontend.Console/Program.fs (hunk 9 above)
Comment (webwarrior-ws) [reply src/GWallet.Frontend.Console/Program.fs]: Done

**Comment by webwarrior-ws** [comment]
Superseded by #295
